### DISCUSSION TITLE
Configure TOTP only for some users

### DISCUSSION THREAD

**Discussion by KramNamez** [keycloak/keycloak#24959] [Q&A] [answered]
We're moving away from TOTPs to Webauthn. However, there are some shared user accounts, for which it would be convenient to be able to have TOTP still available. As far as I can tell, I can only turn off TOTPs entirely, disabling them for all accounts.

Is there a way to disable them, but leave them configured (or configure them again) on a few select accounts?

**Comment by jbman** [comment]
TOTP is a credential stored at each user. If TOTP is configured at a user the default Keycloak Authentication Browser flow will ask for it on login. You can delete it where not needed.

<img src="https://github.com/keycloak/keycloak/assets/789496/4ecf3a95-1aa5-4268-a2a4-d175c84f8b6c" width="400px"></img>

For webauthn you need to adapt the Browser flow.

**Comment by KramNamez** [comment]
So if I turn this option off: 
![image](https://github.com/keycloak/keycloak/assets/108786702/f7e3a16e-79ac-4c0c-9497-a282d0445ed9)

Users can no longer configure one for themselves, but not only will that leave OTPs configured for users who already have them, but allows me to intentionally bypass it where needed?

If so: Do I have to manually go through all users to deactivate it, or how is a migration away from OTPs meant to be handled?

**Reply by danielFesenmeyer** [reply, marked as answer]
Just an idea: You could set a custom user attribute like "totp-disabled" on all affected users, and put these two conditional authenticators before the OTP form:

- Condition - User Configured (makes sure OTP is actually configured for the user)
- Condition - User Attribute (configured to check the custom attribute)

**Reply by KramNamez** [reply]
Hm. That could work - although rather the other way around, right? `totp-enabled` on a few users, only giving _them_ the option to use a TOTP, while for everyone else, it won't work.

My goal is to remove TOTPs from the vast majority of users, after all.

**Reply by danielFesenmeyer** [reply]
Yes, depending on the use case, the other way round.
